### DISCUSSION TITLE
Comprehensive Unguided Scenario-based Exercise - Spring Microservice Design Patterns - Online Banking System

### DISCUSSION THREAD

**Discussion by akash-coded** [akash-coded/spring-framework#173] [Tasks]
### Scenario: Online Banking System – A Comprehensive Exercise

**Context**: Imagine an intricate banking system with the following microservices:

1. **Account Management Service**: Manages user accounts, profile details, account balance, etc.
2. **Transaction Service**: Handles transactions between accounts, including logging and notifying relevant services of transaction events.
3. **Notification Service**: Sends notifications (SMS, Email) to customers regarding account activities.
4. **Audit Service**: Keeps track and logs every action and transaction within the system for auditing purposes.
5. **Fraud Detection Service**: Analyzes transaction patterns, and suspends accounts or blocks transactions that are likely to be fraudulent.

##### Task 1: Account Management and Transaction Services Interaction
- **Scenario**: John wants to transfer money to Emily but encounters a system error after his account is debited.
- **Objective**: Implement a Saga to manage transactions and ensure eventual consistency.
- **Hints**: Emit an event when the transaction is initiated and another when it's completed. If the transaction fails, the saga should initiate a compensating transaction to revert the debit from John's account.
  
##### Task 2: Notification and Fraud Detection Services
- **Scenario**: A high-volume transaction is initiated from John’s account to an overseas account.
- **Objective**: Implement an event-driven communication system where the Fraud Detection Service analyses this unusual transaction, potentially blocking it and emitting an event that triggers the Notification Service.
- **Hints**: Use an event bus/channel. If a suspicious activity is detected, emit a `SuspiciousActivityEvent` which the Notification Service listens for.

##### Task 3: Audit Service Tracking Transactions
- **Scenario**: Every transaction should be audited and logged, providing a trail for future references.
- **Objective**: Implement a Proxy Pattern in the Audit Service that logs transaction details without modifying the Transaction Service.
- **Hints**: Ensure to log relevant transaction details (amount, sender, recipient, timestamp) and possible failure reasons in a separate database.

##### Task 4: API Gateway and External API Consumption
- **Scenario**: A third-party service provides forex rates, and our banking application should present the option to view current rates.
- **Objective**: Implement API Gateway to manage requests to both internal microservices and external APIs.
- **Hints**: The API gateway should route requests to the relevant microservice or external API without exposing the internal architecture to the client application.

##### Task 5: Multi-Service Transaction and Compensation
- **Scenario**: John initiates a transaction which involves currency conversion and transfer.
- **Objective**: Develop a complex saga that manages a multi-service transaction, ensuring each step is successful or initiating compensating transactions.
- **Hints**: If the transaction fails at any point (e.g., forex API is down, insufficient funds after conversion rate applied), implement compensating transactions to revert any preceding transaction steps.

##### Task 6: Implementing Security
- **Scenario**: Secure communication and data storage across microservices, considering sensitive customer data.
- **Objective**: Implement OAuth 2.0 for secure API access and data encryption standards for at-rest and in-transit data.
- **Hints**: Secure all inter-service communication and ensure data is encrypted when stored in databases.

##### Task 7: Real-Time Notification System
- **Scenario**: Real-time account activity alerts for customers.
- **Objective**: Build a WebSocket communication system to notify users in real-time.
- **Hints**: The Notification Service should listen for events (e.g., successful transaction, login) and notify relevant users in real-time via WebSocket.

#### Technical Considerations:

- **API Gateway**: Utilize Zuul or Spring Cloud Gateway for routing requests and isolating internal microservice architectures from client apps.
  
  ```xml
  <!-- For Spring Cloud Gateway -->
  <dependency>
      <groupId>org.springframework.cloud</groupId>
      <artifactId>spring-cloud-starter-gateway</artifactId>
  </dependency>
  ```
  
  Sample routing configuration for API Gateway:
  ```yaml
  spring:
    cloud:
      gateway:
        routes:
        - id: account_service
          uri: lb://account-service
          predicates:
          - Path=/api/account/**
        # Add more routes...
  ```
  
- **Event-Driven System**: Implement events, channels, and queues to manage asynchronous communication between microservices.
- **Data Consistency**: Ensure to manage data consistency across microservices especially in scenarios involving multiple services (e.g., transactions, notifications).
  
#### Task Deliverables:

1. **Microservices’ Source Code**: Well-structured, with thorough inline comments explaining logic and functionality.
2. **API Documentation**: Comprehensive API documentation using Swagger or a similar tool.
3. **Postman Collections**: API testing collections for all the created microservices APIs.

#### Additional Points:

- Explore various failure scenarios and ensure the system can recover or rollback transactions to maintain consistency.
- Implement robust logging across microservices to debug and trace issues and activities.

##### Account Management Service
- **Responsibilities**: Handle CRUD operations for the user account, manage balance.
- **APIs**:
  - `/createAccount`: Create a new user account.
  - `/debit`: Debit the user's account.
  - `/credit`: Credit the user's account.
- **Event Listeners**: 
  - `TransactionFailedEvent`: Listen for failed transactions and compensate by crediting the account.

##### Transaction Service
- **Responsibilities**: Initialize, manage, and finalize transactions between accounts.
- **APIs**:
  - `/initiateTransaction`: Start a new transaction.
  - `/completeTransaction`: Complete a successful transaction.
- **Event Emitters**:
  - Emit `TransactionInitiatedEvent` when a transaction starts.
  - Emit `TransactionCompletedEvent` or `TransactionFailedEvent` based on the outcome.

##### Implementation Details for Task 1
- The Transaction Service should communicate asynchronously with the Account Management Service.
- Upon a transaction request:
  - The Transaction Service emits a `TransactionInitiatedEvent`.
  - Deduct the amount from the sender’s account and emit a `TransactionCompletedEvent` or `TransactionFailedEvent`.
  - If failed, the Account Management Service should listen for `TransactionFailedEvent` and compensate by crediting back the amount.

##### Notification Service
- **Responsibilities**: Listen to various events and send appropriate notifications.
- **Event Listeners**: 
  - `SuspiciousActivityEvent`: Alert the user of potential fraudulent activity.
  - `TransactionCompletedEvent`: Notify the user of a successful transaction.

##### Fraud Detection Service
- **Responsibilities**: Analyze transactions and emit events for suspicious activities.
- **APIs**:
  - `/analyzeTransaction`: Analyze the transaction pattern.
- **Event Emitters**:
  - Emit `SuspiciousActivityEvent` when a potentially fraudulent transaction is detected.

##### Implementation Details for Task 2
- Whenever a transaction is initiated, the Fraud Detection Service analyzes it.
- On detecting suspicious activity, it emits a `SuspiciousActivityEvent`, which the Notification Service listens for.
- Upon receiving `SuspiciousActivityEvent`, the Notification Service alerts the user through SMS or email.

##### Audit Service
- **Responsibilities**: Log every transaction and action within the system.
- **APIs**:
  - `/logActivity`: Log a particular action or transaction.
- **Event Listeners**:
  - `TransactionCompletedEvent`: Log the details of completed transactions.
  - `TransactionFailedEvent`: Log any failed transactions with failure reasons.

##### Implementation Details for Task 3
- Implement the Audit Service in such a way that it transparently logs all transactions without the Transaction Service being aware.
- Utilize the Proxy Pattern to intercept calls to the Transaction Service and log details before forwarding the call.

#### Task 4: API Gateway for Internal and External API Consumption

- **Zuul or Spring Cloud Gateway**: Should route requests to the respective microservice. 

##### Implementation Details for Task 4
- The API Gateway will expose unified endpoints to the client.
- Implement logic to route to internal services or consume external APIs based on the request.

#### Task 5: Multi-Service Transaction with Sagas and Compensation

- **Currency Conversion API**: Integrate an external forex API to get currency rates.
- **Sagas**: Ensure that the saga orchestrates a complex, multi-step transaction.

##### Implementation Details for Task 5
- The saga should orchestrate:
  - Deduction from the sender's account (Account Management Service).
  - Calling the external forex API (via API Gateway) and converting the amount.
  - Crediting the recipient's account (Account Management Service).
- Implement compensating transactions at each step in case of failure.

#### Task 6: Secure Communication and Data

- **OAuth 2.0**: Implement OAuth 2.0 using Spring Security to secure API access.
- **Data Encryption**: Ensure data is encrypted using AES or a similar encryption standard for data at rest.

##### Implementation Details for Task 6
- Set up an Authorization Server to issue tokens for API access.
- Ensure APIs are secured with the necessary scopes and roles.
- Encrypt sensitive data before storage.

#### Task 7: Real-Time Notifications with WebSockets

- **WebSocket Configuration**: Implement WebSocket in Notification Service to push real-time updates to clients.

##### Implementation Details for Task 7
- Implement an event listener in the Notification Service that listens for transaction events and sends real-time updates to connected clients.

### Note:
For each task, ensure to implement robust logging, error handling, and data validation to ensure reliability and maintainability of the services. Add unit tests, integration tests, and API documentation to verify functionality and communicate API usage to other developers.

This scenario combines various patterns and mechanisms in microservices architecture, providing a comprehensive, real-world-like scenario. Implementing it will require intricate coordination between microservices, sagas, proxy patterns, API Gateway mechanisms, and secure data handling, thereby offering a thorough exercise for microservices development practice.

---

##### Task Overview
- **Objective**: Implementing a `Loan Approval Saga` and `Fraud Detection Saga` by embracing an event-driven model.
- **Sub-Tasks**:
    1. **Loan Approval Saga**: When a loan is requested, validate the user's transaction history and approve/deny based on specific criteria.
    2. **Fraud Detection Saga**: Analyze every transaction, if any suspicious transaction is identified (e.g., multiple transactions in a short time), trigger a fraud alert.

##### Implementation Details for Task 11
- **Loan Approval Saga**:
    - When a loan is requested, the `Loan Service` emits a `LoanRequestedEvent`.
    - `Transaction History Service` listens for `LoanRequestedEvent`, validates the user’s transaction history, and emits either `LoanApprovedEvent` or `LoanDeniedEvent` accordingly.
    - `Notification Service` informs the user about the loan approval/denial.
    
- **Fraud Detection Saga**:
    - `Transaction Service` emits a `TransactionInitiatedEvent` for every transaction.
    - `Fraud Detection Service` listens for `TransactionInitiatedEvent`, checks for suspicious patterns, and emits a `FraudAlertEvent` if any fraud is suspected.
    - `Notification Service` informs the user and the admin about the potential fraudulent activity.

##### Code Snippet for Event Emission and Listeners
```java
// Event Emission (In respective services)
applicationEventPublisher.publishEvent(new LoanRequestedEvent(this, loanRequest));
applicationEventPublisher.publishEvent(new TransactionInitiatedEvent(this, transaction));

// Event Listeners (In respective services)
@EventListener
public void onLoanRequestedEvent(LoanRequestedEvent event) {
    // Business logic for event handling
}
```

##### Task Overview
- **Objective**: Implement compensating transactions in the `Loan Approval Saga` to manage failure scenarios.
- **Sub-Tasks**:
    1. If the loan approval fails at any step, initiate a compensating transaction to undo previous transactions.
    2. Inform the user about the failure and any compensatory actions taken.

##### Implementation Details for Task 12
- In the `Loan Approval Saga`, if a failure happens post partial loan approval (e.g., deduction from an account), implement compensating transactions to revert the changes.
- Use events to manage and communicate between microservices.
- Example: If the deduction was successful but the loan process failed later, emit a `LoanFailedEvent` and handle it to revert the deduction.

##### Code Snippet for Compensating Transaction
```java
@EventListener
public void onLoanFailedEvent(LoanFailedEvent event) {
    // Logic for compensating transaction
    transactionService.refundDeduction(event.getTransaction());
}
```

##### Task Overview
- **Objective**: Design an `API Gateway` that acts as a single-entry point and integrates with the `Proxy Pattern` for managing external API calls.
- **Sub-Tasks**:
    - Ensure internal services communicate directly, whereas external service communications are routed via API Gateway using Proxy.
    - The API Gateway, upon receiving a request for external service, invokes the Proxy which manages the request to prevent direct access to actual external service.

##### Implementation Details for Task 13
- Use `Spring Cloud Gateway` to implement API Gateway and use `WebClient` for external API calls.
- Implement a `Proxy Service` which intercepts requests for external APIs, performs additional logic (like caching, logging), and then makes the actual API call.

##### Code Snippet for API Gateway Routing (application.yml)
```yaml
spring:
  cloud:
    gateway:
      routes:
      - id: proxyRoute
        uri: http://localhost:8082  # Proxy Service URL
        predicates:
        - Path=/api/external/**
```

##### Code Snippet for Proxy Service API Call
```java
@Service
public class ProxyService {
    @Autowired
    private WebClient.Builder webClientBuilder;
    
    public String invokeExternalApi(String apiUrl) {
        // Additional Logic before invoking external API
        return webClientBuilder.build()
                .get()
                .uri(apiUrl)
                .retrieve()
                .bodyToMono(String.class)
                .block();
    }
}
```

---

##### Task Overview
- **Objective**: Implement a `Distributed Locking Saga` for concurrently handling multiple user requests for a single bank account operation.
- **Sub-Tasks**:
    1. **Implement a Locking Service**: To manage access to shared resources, primarily bank accounts, across services.
    2. **Transaction Management**: Ensure transaction integrity and reliability amidst concurrent access.

##### Implementation Details for Task 14
- The `Locking Service` listens for `TransactionInitiatedEvent` and locks the corresponding account for transactions.
- If a second transaction is initiated while an account is locked, emit a `TransactionDelayedEvent`.
- Once the initial transaction completes, emit a `TransactionCompletedEvent` to unlock the account and process any delayed transactions.

##### Code Snippet for Event Handling in Locking Service
```java
@EventListener
public void onTransactionInitiatedEvent(TransactionInitiatedEvent event) {
    // Logic to lock account & handle transaction
}

@EventListener
public void onTransactionCompletedEvent(TransactionCompletedEvent event) {
    // Logic to unlock account & handle waiting transactions
}
```

##### Task Overview
- **Objective**: Enhance `Loan Approval Saga` by implementing the `Chain of Responsibility` pattern, decoupling the sender of a request from its receiver by allowing multiple objects to handle the request independently.
- **Sub-Tasks**:
    - Define different criteria as separate handlers in the chain.
    - Each handler decides to either approve, deny, or pass the loan request to the next handler in the chain.

##### Implementation Details for Task 15
- Handlers: `DebtRatioHandler`, `CreditScoreHandler`, `EmploymentStatusHandler`, etc.
- Each handler listens for `LoanRequestedEvent`, applies specific logic, and decides whether to approve, deny, or pass the request to the next handler.
- If the end of the chain is reached with no denial, emit a `LoanApprovedEvent`, otherwise `LoanDeniedEvent`.

##### Code Snippet for a Handler in Chain
```java
public class DebtRatioHandler extends LoanApprovalHandler {

    @Override
    public void handleLoanRequest(LoanRequestedEvent event) {
        if (isDebtRatioAcceptable(event.getUser())) {
            if (nextHandler != null) {
                nextHandler.handleLoanRequest(event);
            } else {
                emitLoanApprovedEvent(event);
            }
        } else {
            emitLoanDeniedEvent(event, "Debt Ratio Unacceptable");
        }
    }
}
```

##### Task Overview
- **Objective**: Implement the `Observer Pattern` in the `Notification Service` to manage alert subscriptions and notify users about different events.
- **Sub-Tasks**:
    - `Subscribe to Events`: Users can subscribe to various notifications like `TransactionAlert`, `LoanStatusAlert`, etc.
    - `Notify Users`: When events are emitted, subscribed users receive the corresponding notifications.

##### Implementation Details for Task 16
- Use Spring Events as a medium to implement the Observer Pattern.
- When events like `TransactionInitiatedEvent`, `LoanApprovedEvent`, etc., are emitted, the `Notification Service` acts as an observer and notifies all subscribed users accordingly.

##### Code Snippet for Observing Events
```java
@EventListener
public void onTransactionInitiatedEvent(TransactionInitiatedEvent event) {
    // Notify all users who are subscribed to TransactionAlert
    notifySubscribedUsers(event, TransactionAlert.class);
}
```

---

# Bonus Tasks

#### Task Overview
- **Objective**: Implement a `Retry Saga` to manage failed transactions due to network issues or service downtimes, ensuring system resilience.
- **Sub-Tasks**:
    1. **Implement a Retry Service**: Should be capable of identifying failed transactions and retrying them after specific intervals.
    2. **Transaction Management**: Safeguard transaction integrity amidst retries.

#### Implementation Details for Task 17
- The `Retry Service` listens for `TransactionFailedEvent` and attempts to retry the transaction after a predefined interval.
- If the transaction fails repeatedly beyond a certain limit, emit a `TransactionPermanentFailedEvent`.

#### Code Snippet for Event Handling in Retry Service
```java
@EventListener
public void onTransactionFailedEvent(TransactionFailedEvent event) {
    // Logic to retry transaction & handle continued failure
}

@EventListener
public void onTransactionPermanentFailedEvent(TransactionPermanentFailedEvent event) {
    // Logic to notify user & log failure
}
```

#### Task Overview
- **Objective**: Implement an `Anti-Corruption Layer Saga` to ensure that communication between services does not corrupt the core business logic of the involved services.
- **Sub-Tasks**:
    - Establish a layer that translates requests and responses between services, safeguarding their respective models.

#### Implementation Details for Task 18
- Create a `Translation Layer` that listens for `ServiceCommunicationEvent`.
- This layer should transform the data from the client service into a form that can be comprehended by the server service and vice-versa.

#### Code Snippet for Anti-Corruption Layer
```java
@EventListener
public void onServiceCommunicationEvent(ServiceCommunicationEvent event) {
    // Logic to translate & adapt service communication
}
```

#### Task Overview
- **Objective**: Implement a `Circuit Breaker Saga` to manage the interaction with external API, providing a fail-fast mechanism.
- **Sub-Tasks**:
    - Monitor API calls and if failure crosses a certain threshold, open the circuit, preventing further calls and fast-failing.
    - Once the service seems stable, close the circuit, allowing calls again.

#### Implementation Details for Task 19
- Implement a service that acts as a Circuit Breaker, listening for `APICallFailedEvent` and `APICallSucceedEvent`.
- Open or close circuit based on logic to protect external API and our services from cascading failures.

#### Code Snippet for Circuit Breaker Service
```java
@EventListener
public void onAPICallFailedEvent(APICallFailedEvent event) {
    // Logic to handle API call failure & potentially open circuit
}

@EventListener
public void onAPICallSucceedEvent(APICallSucceedEvent event) {
    // Logic to handle API call success & potentially close circuit
}
```

### Reflection on Tasks 17-19

These tasks underscore the criticality of managing transactions, service communication, and external API interactions in a microservices architecture, especially within the intricate context of a banking domain. The ability to proficiently navigate through transaction failures, safeguarding service communication, and maintaining system resilience during external API downtimes or failures is pivotal.

These scenarios prop up often in a distributed setup, such as in a banking domain where transaction integrity, communication clarity, and external API stability are crucial. The retry mechanism, anti-corruption layer, and circuit breaker are not just patterns; they’re vital components that shield the system from potential failures and data inconsistencies, ensuring a seamless user experience even when the system encounters failures or disruptions.

Navigating through these sagas provides a multifaceted perspective on managing various aspects of microservices, instilling a profound understanding of implementing and managing distributed systems in a domain as delicate and critical as banking. This hands-on experience will be highly beneficial in grasping the practical application of theoretical concepts and strategies pertinent to microservices architecture.

**Comment by AnuragMishra58** [comment]
Hi @akash-coded 
Please find below link for the above tasks completed so far:-
https://github.com/AnuragMishra58/microservices/blob/main/banking-system/Readme.md
I will update read me with more scenarios

**Comment by arghyagiri** [comment]
@akash-coded : Solution is provided in below repository. More scenario will be added soon and documents will be updated.
https://github.com/arghyagiri/microservice-e2/blob/main/scenario-based-online-banking-system/Read.md#api-test-scripts

**Comment by jay4tech** [comment]
Hi @akash-coded 
Name: Jay Prakash Kumar

Here are possible cases implemented, Some more will be added soon [WIP] - 

https://github.com/jay4tech/account-management-service/tree/msa_173
https://github.com/jay4tech/transaction-service/tree/msa_173
https://github.com/jay4tech/notification-service/tree/msa_173
https://github.com/jay4tech/audit-service/tree/msa_173
https://github.com/jay4tech/fraud-detection-service/tree/msa_173
https://github.com/jay4tech/api-gateway/tree/msa_173

**Comment by shriramram** [comment]
Hi @akash-coded 
Name : Dattatraya Shriram
Description - Please find attached source, still needs to add few scenarios.
[microsevices-banking-system.zip](https://github.com/akash-coded/spring-framework/files/13035811/microsevices-banking-system.zip)

**Comment by Rajashree-Ravi** [comment]
@akash-coded 

Please find attached the implementation for online banking system
[online banking system.zip](https://github.com/akash-coded/spring-framework/files/13037697/online.banking.system.zip)

Github links:

https://github.com/Rajashree-Ravi/banking-api-gateway
https://github.com/Rajashree-Ravi/banking-api-discovery
https://github.com/Rajashree-Ravi/account-service
https://github.com/Rajashree-Ravi/transaction-service
https://github.com/Rajashree-Ravi/notification-service
https://github.com/Rajashree-Ravi/fraud-detection-service
https://github.com/Rajashree-Ravi/audit-service
